easybuild/tools/job/slurm.py

```diff
@@ -0,0 +1,177 @@
+            job_deps_type = JOB_DEPS_TYPE_ABORT_ON_ERROR
+            self.log.info("Using default job dependency type: %s", job_deps_type)
+        else:
+            self.log.info("Using specified job dependency type: %s", job_deps_type)
+
+        if job_deps_type == JOB_DEPS_TYPE_ABORT_ON_ERROR:
+            self.job_deps_type = 'afterok'
+        elif job_deps_type == JOB_DEPS_TYPE_ALWAYS_RUN:
+            self.job_deps_type = 'afterany'
+        else:
+            raise EasyBuildError("Unknown job dependency type specified: %s", job_deps_type)
+
+    def _check_version(self):
+        """Check whether version of Slurm complies with required version."""
+        (out, _) = run_cmd("sbatch --version", trace=False)
+        slurm_ver = out.strip().split(' ')[-1]
+        self.log.info("Found Slurm version %s", slurm_ver)
+
+        if LooseVersion(slurm_ver) < LooseVersion(self.REQ_VERSION):
+            raise EasyBuildError("Found Slurm version %s, but version %s or more recent is required",
+                                 slurm_ver, self.REQ_VERSION)
+
+    def init(self):
+        """
+        Initialise the PySlurm job backend.
+        """
+        self._submitted = []
+
+    def queue(self, job, dependencies=frozenset()):
+        """
+        Add a job to the queue.
+
+        :param dependencies: jobs on which this job depends.
+        """
+        submit_cmd = 'sbatch'
+
+        if dependencies:
+            job.job_specs['dependency'] = self.job_deps_type + ':' + ':'.join(str(d.jobid) for d in dependencies)
+            # make sure job that has invalid dependencies doesn't remain queued indefinitely
+            submit_cmd += " --kill-on-invalid-dep=yes"
+
+        # submit job with hold in place
+        job.job_specs['hold'] = True
+
+        self.log.info("Submitting job with following specs: %s", job.job_specs)
+        for key in sorted(job.job_specs):
+            if key in ['hold']:
+                if job.job_specs[key]:
+                    submit_cmd += " --%s" % key
+            else:
+                submit_cmd += ' --%s "%s"' % (key, job.job_specs[key])
+
+        (out, _) = run_cmd(submit_cmd, trace=False)
+
+        jobid_regex = re.compile("^Submitted batch job (?P<jobid>[0-9]+)")
+
+        res = jobid_regex.search(out)
+        if res:
+            job.jobid = res.group('jobid')
+            self.log.info("Job submitted, got job ID %s", job.jobid)
+        else:
+            raise EasyBuildError("Failed to determine job ID from output of submission command: %s", out)
+
+        self._submitted.append(job)
+
+    def complete(self):
+        """
+        Complete a bulk job submission.
+
+        Release all user holds on submitted jobs, and disconnect from server.
+        """
+        for job in self._submitted:
+            if job.job_specs['hold']:
+                self.log.info("releasing user hold on job %s" % job.jobid)
+                run_cmd("scontrol release %s" % job.jobid, trace=False)
```

Comment:
scontrol release can take multiple jobids, line them up and call run_cmd once instead.

Reply:
done